virttest/qemu_devices/qcontainer.py

```diff
@@ -3293,11 +3293,7 @@ def thread_context_define_by_params(self, params, name):
         Create thread-context object from params.
         """
         tc_params = Params()
-        prefix = "vm_thread_context_"
-        for key in params:
-            if key.startswith(prefix):
-                new_key = key.rsplit(prefix)[1]
-                tc_params[new_key] = params[key]
+        tc_params["node-affinity"] = params["vm_thread_context_node-affinity"]
```

Comment:
Is this parameter always required? If it's not defined, the case will be canceled.

`thread-context` only has 2 options
```
# /usr/libexec/qemu-kvm -object thread-context,?
thread-context options:
  cpu-affinity=<int>
  node-affinity=<int>
```

so a possible way is to maintain a option list and check if it's in the params, but if qemu extend the option support, we have to update it accordingly.

Another way I can think of is to check for suffixes should not in params.objects('vm_thread_contexts'), like the following:
```
>>> params = utils_params.Params({"vm_thread_contexts": "tc1 tc2", "vm_thread_context_foo_tc1": "bar1", "vm_thread_context_foo_tc2": "bar2"})
>>> thread_contexts = params.objects("vm_thread_contexts")
>>> tc1_params = params.object_params('tc1')
>>> tc1_params
{'vm_thread_contexts': 'tc1 tc2', 'vm_thread_context_foo_tc1': 'bar1', 'vm_thread_context_foo_tc2': 'bar2', 'vm_thread_context_foo': 'bar1'}
>>>
```

So besides checking the prefix, we can also check the suffix should not in `params.objects("vm_thread_contexts")`.

What do you think?

Reply:
BTW, if you perfect the current fix, maybe `copy_from_keys` can be used, `tc_params` can use `params.copy_from_keys(options)`, even if the key is not defined, it will be skipped. the `options` can be maintained in the test case side or in this method, it can be dynamic.

Reply:
> BTW, if you perfect the current fix, maybe `copy_from_keys` can be used, `tc_params` can use `params.copy_from_keys(options)`, even if the key is not defined, it will be skipped. the `options` can be maintained in the test case side or in this method, it can be dynamic.

@PaulYuuu so this latter idea will be like?:
```
options = ["cpu-affinity", "node-affinity"]
params = params.copy_from_keys(options)
dev = qdevices.QObject("thread-context", params=params)
```

Reply:
To make sure it's safer, we can also add the prefix just like @luckyh mentioned above.

Reply:
> To make sure it's safer, we can also add the prefix just like @luckyh mentioned above.

@PaulYuuu I think I don't get your point... then, how do I transform `vm_thread_context_node-affinity` into `node-affinity` using the above suggestion? IIUC the `copy_from_keys` function only copies the ones matching the options...

Reply:
Firstly, I considered we should add the prefix, like:
```
# cfg
vm_thread_context_options = "vm_thread_context_cpu-affinity vm_thread_context_node-affinity"
```

```
# py
params = params.copy_from_keys(params.get_list("vm_thread_context_options"))
```

But as we already have object_params before defining thread-context, so I think it's okay to use such as `cpu-affinity_tc1` to define the device.

Reply:
> py
> params = params.copy_from_keys(params.get_list("vm_thread_context_options"))

@PaulYuuu 

This isn't just the same as params.objects functionality, which currently is not working fine? Anyways defining the `vm_thread_context_options` in the test cfg doesn't work, it returns an empty list at `qcontainer.py`

Reply:
@PaulYuuu as I discussed previously with @luckyh and I was trying to explain is: the cartesian configuration is not working properly when doing params.object_params(), keeps all the params not doing any kind of filtering... anyways, I've implemented the second option skipping this function and with a little modification (I am going to send a tp-qemu patch) all of them passed
```
 (1/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_prealloc_threads.q35: STARTED
 (1/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_prealloc_threads.q35: PASS (177.17 s)
 (2/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_prealloc_threads.sandbox_off.q35: STARTED
 (2/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_prealloc_threads.sandbox_off.q35: PASS (178.03 s)
 (3/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_prealloc_threads.boot_cpu_affinity.q35: STARTED
 (3/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_prealloc_threads.boot_cpu_affinity.q35: PASS (177.78 s)
 (4/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_prealloc_threads.boot_cpu_affinity.sandbox_off.q35: STARTED
 (4/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_prealloc_threads.boot_cpu_affinity.sandbox_off.q35: PASS (179.20 s)
 (5/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_node_affinity.q35: STARTED
 (5/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_node_affinity.q35: PASS (40.41 s)
 (6/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_prealloc_handling.q35: STARTED
 (6/6) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.numa_prealloc_handling.q35: PASS (7.09 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```